sql/walrus--0.1.sql

```diff
@@ -290,182 +285,214 @@ declare
-        'table', wal ->> 'table',
-        'type', action,
-        'commit_timestamp', (wal ->> 'timestamp')::text::timestamp with time zone,
-        'columns', (
-            select
-                jsonb_agg(
-                    jsonb_build_object(
-                        'name', pa.attname,
-                        'type', pt.typname
-                    )
-                    order by pa.attnum asc
-                )
-            from
-                pg_attribute pa
-                join pg_type pt
-                    on pa.atttypid = pt.oid
-            where
-                attrelid = entity_
-                and attnum > 0
-                and pg_catalog.has_column_privilege('authenticated', entity_, pa.attname, 'SELECT')
-        )
-    )
-    -- Add "record" key for insert and update
-    || case
-        when error_record_exceeds_max_size then jsonb_build_object('record', '{}'::jsonb)
-        when action in ('INSERT', 'UPDATE') then
-            jsonb_build_object(
-                'record',
-                (select jsonb_object_agg((c).name, (c).value) from unnest(columns) c where (c).is_selectable)
+    for working_role in select * from unnest(roles) loop
+
+        -- Update `is_selectable` for columns and old_columns
+        columns =
+            array_agg(
+                (
+                    c.name,
+                    c.type,
+                    c.value,
+                    c.is_pkey,
+                    pg_catalog.has_column_privilege(working_role, entity_, c.name, 'SELECT')
+                )::realtime.wal_column
             )
-        else '{}'::jsonb
-    end
-    -- Add "old_record" key for update and delete
-    || case
-        when error_record_exceeds_max_size then jsonb_build_object('old_record', '{}'::jsonb)
-        when action in ('UPDATE', 'DELETE') then
-            jsonb_build_object(
-                'old_record',
-                (select jsonb_object_agg((c).name, (c).value) from unnest(old_columns) c where (c).is_selectable)
+            from
+                unnest(columns) c;
+
+        old_columns =
+                array_agg(
+                    (
+                        c.name,
+                        c.type,
+                        c.value,
+                        c.is_pkey,
+                        pg_catalog.has_column_privilege(working_role, entity_, c.name, 'SELECT')
+                    )::realtime.wal_column
+                )
+                from
+                    unnest(old_columns) c;
+
+        if action <> 'DELETE' and count(1) = 0 from unnest(columns) c where c.is_pkey then
+            return next (
+                null,
+                is_rls_enabled,
+                (select array_agg(id) from realtime.subscription where entity = entity_ and claims_role = working_role),
+                array['Error 400: Bad Request, no primary key']
+            )::realtime.wal_rls;
+
+        -- The claims role does not have SELECT permission to the primary key of entity
```

Comment:
@olirice i think you're only checking select permission for primary key b/c it's more performant and check for the actual column is done in `realtime.subscription_check_filters`? 

Realtime server still sprays every change to every filter so we won't have great column security even after these changes. After this PR is in we should talk about better filter security.

Reply:
> why are you only checking select permission for primary key?

we only check the primary key here because the primary key cols are all that is necessary to check if the RLS passes.

- When a subscription is created we check to make sure that the subscriptions `claim_role` has access to the columns referenced by filters https://github.com/supabase/walrus/blob/generic_claims/sql/walrus--0.1.sql#L82
- When we're processing the WAL, we strip the columns the role shouldn't see from `record` and `old_record` keys https://github.com/supabase/walrus/blob/generic_claims/sql/walrus--0.1.sql#L424